Issue 366 push pull ui improvements

ggalmazor opened the pull request:
Closes #366 
This PR follows the work started by @mayank8318 in #405

This PR also migrates the Pull and Push tabs' UI layout to IntelliJ's GUI Designer.

The new tab implementations mirror the same structures and abstractions already in place for the export tab.

The new Aggregate connection dialog handles better any error scenarios produced by wrong user input or unexpected server responses. Any error will be gracefully reported to the user, letting her accept the configuration only after all necessary checks have been made and data has been validated.

During the development of this PR, I detected some problems with the way the Export tab updates its forms table when new forms are pulled. Since I had to change things to make it work on the push tab anyway, I fixed that on the export tab as well.

I had to change the way the pull and push tabs reacted to events because they shared a common transfer event classes. This made the push tab incorrectly react to events generated from the pull tab and viceversa.

#### What has been done to verify that this works as intended?
Run automated tests and tested the changes manually to the UI and the CLI. 
- When testing pulling/pushing using an Aggregate server as a source, tested it both with and without credentials (changing security configuration on the server to allow/forbid listing and uploading for the anonymous user)
- Also tested the new Aggregate connection dialog with wrong URLs, invalid URLs and empty fields to test the different error outcomes.
- Tested to pull forms from an Android device through a USB connection.

This PR will require a full QA sweep before releasing, though.

#### Why is this the best possible solution? Were any other approaches considered?
We could have made a small fix on the previous connection testing code and UI layouts, but I discarded that for the following reasons:
- The HTTP interaction code relies on global state and side effects and changing that could have potentially affected all other operations, making this PR riskier.
- Duplicating and modifying the previous HTTP interaction code was an option but taking the chance for improving the design and getting rid of the global state and side-effects made more strategic sense.
- Changing the previous UI layout would have been very difficult (we also had the issue of vertically aligning the progress bar, and decided not to fix it for the same reason)
- Migrating the layout to IntelliJ's GUI Designer was long overdue anyway

#### Are there any risks to merging this code? If so, what are they?
I can't think of any.

#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.
Nope.

ggalmazor commented:
This is missing some stuff. Please, don't review yet

ggalmazor commented:
@yanokwa, @lognaturel, I think this one is ready for review now.

yanokwa commented:
@ggalmazor  Did you get a chance to test with local file systems?

I also seem to have an issue where if I auth against the server successfully, and put in bad credentials, I can still get a form list.

ggalmazor commented:
> Did you get a chance to test with local file systems?

Yeah, I used an Android phone connected to my computer with an USB cable, selected the odk folder, and pulled some forms successfully.

> Also, I seem to have an issue where if I auth against the server successfully, and put in bad credentials, I can still get a form list.

If the anonymous user has enough priviledges to list and upload, it doesn't matter that you send wrong credentials, because Aggregate ignores that. Briefcase can only react to an HTTP 401 response from the server. In this case, you can leave the credentials empty.

I could add an extra check for when the user adds credentials (maybe sending them to the login servlet) but I don't know if artificially failing something that would otherwise work makes sense. What do you think, @yanokwa?

dcbriccetti commented on src/org/opendatakit/briefcase/ui/reused/source/RemoteServerDialog.java:
```diff
@@ -0,0 +1,72 @@
+        form.hideDialog();
+      } else {
+        String error = response.isRedirection()
+            ? "Redirection detected.\n\nPlease, review the connection parameters and try again."
```
Better without commas after these “please”es.

dcbriccetti commented on src/org/opendatakit/briefcase/ui/reused/source/Source.java:
```diff
@@ -0,0 +1,197 @@
+import org.opendatakit.briefcase.util.FileSystemUtils;
+import org.opendatakit.briefcase.util.TransferAction;
+
+public interface Source<T> {
```
Doc comment?

dcbriccetti commented on src/org/opendatakit/briefcase/operations/ClearPreferences.java:
```diff
@@ -37,6 +39,8 @@ private static void clear() {
     Stream.of(
         PullTransferPanel.class,
         PushTransferPanel.class,
+        PullPanel.class,
```
I like this use of Stream

dcbriccetti commented:
Connection refused causes hang.
Exception in thread "AWT-EventQueue-0" java.io.UncheckedIOException: org.apache.http.conn.HttpHostConnectException: Connect to localhost:8080 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1, localhost/fe80:0:0:0:0:0:0:1%1] failed: Connection refused (Connection refused)

dcbriccetti commented:
It seems I have to configure Aggregate once each for both push _and_ pull.

dcbriccetti commented:
This may have nothing to do with this PR, but ...
“Success” is not a sentence and shouldn’t have a period. And wouldn’t “Pulled” be a better word for a pull status?
![screen shot 2018-05-22 at 9 12 00 pm](https://user-images.githubusercontent.com/55099/40403171-fe19cb98-5e04-11e8-8b72-dd26131defad.png)

dcbriccetti commented on src/org/opendatakit/briefcase/ui/pull/components/PullFormsTableView.java:
```diff
@@ -0,0 +1,113 @@
+      button.setBackground(isSelected ? table.getSelectionBackground() : table.getBackground());
+      return button;
+    };
+  }
```
I’d like your opinion on this alternative. Is it simpler or clearer?

![screen shot 2018-05-22 at 9 34 08 pm](https://user-images.githubusercontent.com/55099/40403671-27a4864e-5e08-11e8-9387-1e5e935bd62e.png)

ggalmazor replied:
That is much much much better! Thanks

ggalmazor commented:
> It seems I have to configure Aggregate once each for both push and pull.

That is correct. Sources of each tab are managed independently

ggalmazor commented:
> Connection refused causes hang.
> Exception in thread "AWT-EventQueue-0" java.io.UncheckedIOException: org.apache.http.conn.HttpHostConnectException: Connect to localhost:8080 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1, localhost/fe80:0:0:0:0:0:0:1%1] failed: Connection refused (Connection refused)
> 
> Timeout causes a hang.
> Exception in thread "AWT-EventQueue-0" java.io.UncheckedIOException: org.apache.http.conn.HttpHostConnectException: Connect to 192.168.1.50:80 [/192.168.1.50] failed: Operation timed out (Connection timed out)

Could you copy the stacktrace, please? Also, how did you produce the refused connection scenario? (I think I know how to reproduce the timeout)

ggalmazor commented:
@dcbriccetti, I think I've solved all the issues now.

I've added some code to handle connection problems and I've added a progress bar to the dialog to give the user feedback should the connection test go longer than expected.

**I've set the timeout to 10 secs**. Do you think this value is OK, @yanokwa?

dcbriccetti commented on src/org/opendatakit/briefcase/reused/OptionalProduct.java:
```diff
@@ -0,0 +1,128 @@
+import java.util.Optional;
+
+/**
+ * This class represents the type-safe Application of an arbitrary amount of {@link Optional} values.
```
For discrete measurements, use number instead of amount. What amount of milk do you want? How many games did you win.

ggalmazor replied:
Thanks for the explanation :)

dcbriccetti commented on src/org/opendatakit/briefcase/reused/OptionalProduct.java:
```diff
@@ -0,0 +1,128 @@
+ * <p>
+ * When any of its members is missing, it behaves like a {@link Optional#empty()}
+ */
+public interface OptionalProduct {
```
Are there libraries for Java that we could be using instead of writing all this?

ggalmazor replied:
I only know https://arrow-kt.io/docs/typeclasses/applicative/

lognaturel commented:
A 10 second time-out seems good to me @ggalmazor.

@dcbriccetti thanks so much for your careful look through. The review is more to see if there are any opportunities to solve problems more efficiently or to e.g. leverage libs to help and you're well positioned to identify those things.

Agreed QA will be super important here. I'm wondering whether it might be more helpful to merge some of these big PRs and do a thorough QA pass and file issues for any problems that emerge rather than doing QA for each PR. What do you think @kkrawczyk123 @ggalmazor @dcbriccetti?

dcbriccetti commented:
Merge and QA seems good.--

Sent from a hand computer

ggalmazor commented:
Closing in favor of #474

kkrawczyk123 commented:
@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"
